File: Baekjoon/DynamicProgramming/[14002]가장 긴 증가하는 부분 수열 4/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

#define endl "\n"

vector<int> seq, dp, lis;
int N;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> N;

  seq = vector<int>(N);
  dp = vector<int>(N, 1);

  for (int i = 0; i < N; i++) {
    cin >> seq[i];
  }

  for (int i = 1; i < N; i++) {
    for (int j = i - 1; j >= 0; j--) {
      if (seq[i] > seq[j]) {
        dp[i] = max(dp[i], dp[j] + 1);
      }
    }
  }

  int target = *max_element(dp.begin(), dp.end());

  lis = vector<int>(target);

  for (int i = N - 1; i >= 0; i--) {
    if (dp[i] == target) {
      lis[target - 1] = seq[i];
      target--;
    }
  }

  cout << lis.size() << endl;

  for (int i = 0; i < lis.size(); i++) {
    cout << lis[i] << " ";
  }
  cout << endl;

  return 0;
}

File: Baekjoon/DynamicProgramming/[14002]가장 긴 증가하는 부분 수열 4/solution.py
import sys

N = int(sys.stdin.readline())
seq = list(map(int, sys.stdin.readline().split()))
dp = [1 for i in range(N)]

for i in range(1, N):
    for j in range(i - 1, -1, -1):
        if seq[i] > seq[j]:
            dp[i] = max(dp[i], dp[j] + 1)

target = max(dp)
lis = [0 for i in range(target)]

for i in range(N - 1, -1, -1):
    if dp[i] == target:
        lis[target - 1] = seq[i]
        target -= 1

sys.stdout.write(str(len(lis)) + "\n")

for n in lis:
    sys.stdout.write(str(n) + " ")
sys.stdout.write("\n")
